src/main/java/org/openstreetmap/atlas/geography/atlas/dynamic/DynamicAtlasExpander.java

```diff
@@ -0,0 +1,691 @@
+            buildUnderlyingMultiAtlas();
+            // Record the current list of shards
+            currentShards = new HashSet<>(this.loadedShards.keySet());
+            // Loop through the entities again to find potential shards to add.
+            browseForPotentialNewShards();
+            browseForPotentialNewShardsFromAggressiveRelations();
+        }
+        this.preemptiveLoadDone = true;
+    }
+
+    boolean relationCovered(final Relation relation)
+    {
+        final Set<Long> parentRelationIdentifierTree = new HashSet<>();
+        parentRelationIdentifierTree.add(relation.getIdentifier());
+        return relationCoveredInternal(relation, parentRelationIdentifierTree);
+    }
+
+    private void addNewShardLog(final Shard shard)
+    {
+        if (logger.isInfoEnabled())
+        {
+            final Atlas loaded = this.loadedShards.get(shard);
+            if (loaded == null)
+            {
+                logger.info("{}: Loading new shard {} found no new Atlas.",
+                        this.dynamicAtlas.getName(), shard.getName());
+            }
+            else
+            {
+                logger.info("{}: Loading new shard {} found a new Atlas {} of size {}",
+                        this.dynamicAtlas.getName(), shard.getName(), loaded.getName(),
+                        loaded.size());
+            }
+        }
+    }
+
+    private void addNewShards(final Iterable<? extends Shard> shards)
+    {
+        final Set<Shard> initialNonEmptyLoadedShards = nonNullShards();
+        for (final Shard shard : shards)
+        {
+            if (!this.loadedShards.containsKey(shard))
+            {
+                this.loadedShards.put(shard, this.atlasFetcher.apply(shard).orElse(null));
+                addNewShardLog(shard);
+            }
+        }
+        final List<Atlas> nonNullAtlasShards = getNonNullAtlasShards();
+        if (!nonNullAtlasShards.isEmpty())
+        {
+            if (shouldBuildUnderlyingMultiAtlasWhenAddingNewShards(initialNonEmptyLoadedShards))
+            {
+                // Load the new current atlas only if it is the first time, or it is not the
+                // first time, and the policy is not to defer loading.
+                buildUnderlyingMultiAtlas();
+            }
+        }
+        else
+        {
+            // There should always be a non-null atlas in that list, coming from the initial Shard.
+            throw new CoreException("{}: There is no data to load for initial shard!",
+                    this.dynamicAtlas.getName());
+        }
+    }
+
+    private boolean areaCoversInitialShardBounds(final Area area)
+    {
+        return this.policy.getInitialShardsBounds().overlaps(area.asPolygon());
+    }
+
+    private void browseForPotentialNewShards()
+    {
+        // Look at regular entities
+        this.dynamicAtlas.entities();
+    }
+
+    private void browseForPotentialNewShardsFromAggressiveRelations()
```

Comment:
This new method makes sure to browse for new shards and relations only once, and not for every relation considered for expansion.